src/Features/Core/Portable/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,337 @@
+                        {
+                            _symbolValueUsageStateMap.Add(symbolContext.Symbol, ValueUsageInfo.None);
+                        }
+                    }
+                }
+            }
+
+            public void AnalyzeFieldInitializer(OperationAnalysisContext operationContext)
+            {
+                var initializer = (IFieldInitializerOperation)operationContext.Operation;
+                if (!initializer.Value.ConstantValue.HasValue)
+                {
+                    foreach (var field in initializer.InitializedFields)
+                    {
+                        OnSymbolUsage(field, ValueUsageInfo.Write);
+                    }
+                }
+            }
+
+            private void OnSymbolUsage(ISymbol memberSymbol, ValueUsageInfo usageInfo)
+            {
+                Debug.Assert(IsCandidateSymbol(memberSymbol));
+
+                lock (_gate)
+                {
+                    // Update the usage info for the memberSymbol
+                    if (_symbolValueUsageStateMap.TryGetValue(memberSymbol, out var currentUsageInfo))
+                    {
+                        usageInfo = currentUsageInfo | usageInfo;
+                    }
+
+                    _symbolValueUsageStateMap[memberSymbol] = usageInfo;
+                }
+            }
+
+            private bool TryRemove(ISymbol memberSymbol, out ValueUsageInfo valueUsageInfo)
+            {
+                lock (_gate)
+                {
+                    if (_symbolValueUsageStateMap.TryGetValue(memberSymbol, out valueUsageInfo))
+                    {
+                        _symbolValueUsageStateMap.Remove(memberSymbol);
+                        return true;
+                    }
+
+                    return false;
+                }
+            }
+
+            public void AnalyzeMemberReferenceOperation(OperationAnalysisContext operationContext)
+            {
+                var memberReference = (IMemberReferenceOperation)operationContext.Operation;
+                if (IsCandidateSymbol(memberReference.Member))
+                {
+                    // Get the value usage info.
+                    var valueUsageInfo = memberReference.GetValueUsageInfo();
+
+                    if (valueUsageInfo == ValueUsageInfo.ReadWrite)
+                    {
+                        Debug.Assert(memberReference.Parent is ICompoundAssignmentOperation compoundAssignment &&
+                            compoundAssignment.Target == memberReference ||
+                            memberReference.Parent is IIncrementOrDecrementOperation);
+
+                        // Compound assignment or increment whose value is being dropped (parent has null type)
+                        // is treated as a Write as the value was never actually 'read' in a way that is observable.
+                        if (memberReference.Parent.Parent?.Type == null)
```

Comment:
these are worth showing examples of the difference.  it's clear to me since i brought up the use case as part of the PR.  but someone else reading this would not really know what this was referring to.

Reply:
Sure, done.

---
In reply to: [213416141](https://github.com/dotnet/roslyn/pull/29511#discussion_r213416141) [](ancestors = 213416141)